### ISSUE TITLE
Many tests fail: error: couldn't load image 'texture.png'

### ISSUE THREAD

**Issue by yurivict** [akb825/Cuttlefish#19]
```
/bin/ln -s /usr/ports/misc/cuttlefish/work/Cuttlefish-2.7.2/tool/test/texture.png /usr/ports/misc/cuttlefish/work/.build/ || true
ln: /usr/ports/misc/cuttlefish/work/.build//texture.png: File exists
GTest not installed. Skipping tests.
PVRTexLib not found.
-- Configuring done (0.1s)
-- Generating done (0.0s)
-- Build files have been written to: /usr/ports/misc/cuttlefish/work/.build
ninja: no work to do.
[  0% 1/1] cd /usr/ports/misc/cuttlefish/work/.build && /usr/local/bin/ctest --force-new-ctest-process
Test project /usr/ports/misc/cuttlefish/work/.build
      Start  1: CuttlefishNoArgs
 1/83 Test  #1: CuttlefishNoArgs .....................   Passed    0.01 sec
      Start  2: CuttlefishHelp
 2/83 Test  #2: CuttlefishHelp .......................   Passed    0.01 sec
      Start  3: CuttlefishSaveDds
 3/83 Test  #3: CuttlefishSaveDds ....................***Failed    0.01 sec
error: couldn't load image 'texture.png'
got exit code 2; expected 0
exit: Illegal number: -1

      Start  4: CuttlefishSaveKtx
 4/83 Test  #4: CuttlefishSaveKtx ....................***Failed    0.01 sec
error: couldn't load image 'texture.png'
got exit code 2; expected 0
exit: Illegal number: -1

      Start  5: CuttlefishSavePvr
 5/83 Test  #5: CuttlefishSavePvr ....................***Failed    0.01 sec
error: couldn't load image 'texture.png'
got exit code 2; expected 0
exit: Illegal number: -1

      Start  6: CuttlefishSaveUnknownFileType
 6/83 Test  #6: CuttlefishSaveUnknownFileType ........   Passed    0.01 sec
      Start  7: CuttlefishSaveMissingFileTypeParam
 7/83 Test  #7: CuttlefishSaveMissingFileTypeParam ...   Passed    0.01 sec
      Start  8: CuttlefishMissingOutput
 8/83 Test  #8: CuttlefishMissingOutput ..............   Passed    0.01 sec
      Start  9: CuttlefishMultipleOutput
 9/83 Test  #9: CuttlefishMultipleOutput .............   Passed    0.01 sec
      Start 10: CuttlefishSaveBC3
10/83 Test #10: CuttlefishSaveBC3 ....................***Failed    0.01 sec
error: couldn't load image 'texture.png'
got exit code 2; expected 0
exit: Illegal number: -1

```

The project is built out-source, in  ```/usr/ports/misc/cuttlefish/work/.build```
There is a symbolic link that the port creates there for 'texture.png' but this error occurs any way.

System call log also contains this error:
```
 46879 ctest    GIO   fd 3 wrote 212 bytes
       "Command: "/usr/ports/misc/cuttlefish/work/Cuttlefish-2.7.2/tool/test/run-test.sh" "/usr/ports/misc/cuttlefish/work/.build/output/cuttlefish" "-i texture.png -f R8G8B8A8 -s rgb -o /dev/null --file-format pvr" "1"
       "
 46879 ctest    RET   write 212/0xd4
 46879 ctest    CALL  write(0x3,0x82f60e840,0x46)
 46879 ctest    GIO   fd 3 wrote 70 bytes
       "Directory: /usr/ports/misc/cuttlefish/work/Cuttlefish-2.7.2/tool/test
       "
 46879 ctest    RET   write 70/0x46
 46879 ctest    CALL  write(0x3,0x82f60e840,0x3e)
 46879 ctest    GIO   fd 3 wrote 62 bytes
       ""CuttlefishSwizzleInvalidNumber" start time: May 31 08:24 PDT
       "
 46879 ctest    RET   write 62/0x3e
 46879 ctest    CALL  write(0x3,0x82f60e840,0x8)
 46879 ctest    GIO   fd 3 wrote 8 bytes
       "Output:
       "
 46879 ctest    RET   write 8
 46879 ctest    CALL  write(0x3,0x82f60e840,0x3b)
 46879 ctest    GIO   fd 3 wrote 59 bytes
       "----------------------------------------------------------
       "
 46879 ctest    RET   write 59/0x3b
 46879 ctest    CALL  write(0x3,0x82f60e840,0x2467)
 46879 ctest    GIO   fd 3 wrote 4096 bytes
       "error: swizzle must contain 4 elements of r, g, b, a, or x

        Usage: cuttlefish [options]

        General options:
          -h, --help     display this help message
          -j, --jobs [n] the number of jobs to convert with; if n is omitted, use all
                         available cores
          -q, --quiet    suppress all non-error output
          -v, --verbose  verbose output

        Input files: at least one required, cannot mix different types
          -i, --input file               the input image for a standard texture
          -a, --array [n] file           the input image for an array or 3D texture:
                                           n: the index of the image
                                           file: the path to the image
          -c, --cube face file           the input image for a cube map:
                                           face: the face, which is one of:
                                             +x, -x, +y, -y, +z, -z
                                           file: the path to the image
          -C, --cube-array n face file   the input image for a cube map:
                                           n: the index of the image
                                           face: the face, which is one of:
                                             +x, -x, +y, -y, +z, -z
                                           file: the path to the image
          -I, --input-list type file     specify a file with a list of image paths:
                                           type: type of texture, which is one of the
                                             following: image, array, cube, cube-array;
                                             cube faces are in the order:
                                             +x, -x, +y, -y, +z, -z
                                           file: path to a file containing a list of
```

Version: 2.7.2
FreeBSD 13.2

**Comment by akb825** [comment]
@yurivict the tests that are failing are defined in `tool/CMakeLists.txt`. They are configured to use the `tool/test` directory as the working directory for the tests themselves. (i.e. equivalent to switching to that directory before running the test) There is no need to create a symlink, and the symlink you created will be unused.

You can verify that the working directory is set properly by viewing the file `/usr/ports/misc/cuttlefish/work/.build/tool/CTestTestfile.cmake`, which is generated when running `cmake` and will use absolute paths. Based on your above paths, I expect this to be the first pair of lines after the comment header:
```
add_test(CuttlefishNoArgs "/usr/ports/misc/cuttlefish/work/Cuttlefish-2.7.2/tool/test/run-test.sh" "/usr/ports/misc/cuttlefish/work/.build/output/cuttlefish" "" "1")
set_tests_properties(CuttlefishNoArgs PROPERTIES  WORKING_DIRECTORY "/usr/ports/misc/cuttlefish/work/Cuttlefish-2.7.2/tool/test" _BACKTRACE_TRIPLES "/usr/ports/misc/cuttlefish/work/Cuttlefish-2.7.2/tool/CMakeLists.txt;21;add_test;/usr/ports/misc/cuttlefish/work/Cuttlefish-2.7.2/tool/CMakeLists.txt;0;")
```

As a sanity check, you should also make sure that `/usr/ports/misc/cuttlefish/work/Cuttlefish-2.7.2/tool/test/texture.png` exists and has the appropriate read permissions. To verify that the contents are correct, this is the output for `md5sum`:
```
9ec29200d7d5db889f22afca708f55d0  tool/test/texture.png
```

Another sanity check you can try is to edit `tool/test/run-test.sh` to add a call to `pwd` and `ls` before the line `"$command" $args` to verify that the working directory and directory contents are correct.

I tried to reproduce your build setup as much as I could locally without success. I'm currently under Linux as I don't have a BSD VM handy at the moment, though I did run the tests successfully in BSD earlier when you reported a similar issue for 2.7.1.

> System call log also contains this error:

This is expected, as the `CuttlefishSwizzleInvalidChannel` test case intentionally passes an invalid swizzle and expects the tool to fail.

A couple other notes:
* It looks like it's not building the unit tests as gtest (aka googletest) isn't available. The tests that are running and having failure are tests against the CLI tool, mainly to verify the command-line arguments, but unit tests are arguably the most important to verify the internals are working correctly.
* The port currently lists ISPC as a test dependency, it should be an optional build dependency. It's used to utilize the best quality and fastest texture encoders, though fallback encoders will be used if ISPC isn't available. ISPC supports x86 and ARM targets, while other architectures like PPC would have to use the fallback encoders as ISPC won't be supported.

**Comment by akb825** [comment]
@yurivict I have updated my FreeBSD VM to 13.2 and done the following steps to attempt to recreate your situation:
* Created a `.build` folder as a sibling directory to the Cuttlefish checkout. (at the 2.7.2 release)
* Ran CMake with the `ninja` generator and built the code.
* Ran `ctest --force-new-ctest-process` from the `.build` directory.
* I tried both with and without the googletest package installed to ensure I had the same overall build configuration.

All tests succeeded for me when running with the above steps. Any extra information based on my suggestions in my previous comment would be helpful to figure out why the tests are unable to find `texture.png` for your build.

**Comment by akb825** [comment]
@yurivict checking in again to see if you have any more info. As described above, I tried to emulate your commands as close as I could within FreeBSD VM and still haven't been able to reproduce the issue.

You've reported similar issues a couple times before, I'd like to track down the root cause for what's different either in your environment or how you're preparing and performing your build. To do so I need either:
1. Some extra debugging information as described in my first comment, such as:
    * Verification that the contents of `/usr/ports/misc/cuttlefish/work/.build/tool/CTestTestfile.cmake` is as expected. (see my first comment)
    * Verification that `/usr/ports/misc/cuttlefish/work/Cuttlefish-2.7.2/tool/test/texture.png` exists, has the appropriate permissions, and has the proper contents.
    * Checks within the `/usr/ports/misc/cuttlefish/work/Cuttlefish-2.7.2/tool/test/run-test.sh` script that the current working directly is correct and is able to see `texture.png` from that directory. (see my first comment)
2. Clear instructions to reproduce in a clean environment. Checking out and building the project on FreeBSD is not enough to reproduce the failures.

**Comment by yurivict** [comment]
Hi @akb825 ,

If you already have the FreeBSD VM - the easiest way to reproduce is:
1. cd /usr/ports/misc/cuttlefish
2. pkg install -A `make missing`
3. make test

**Comment by akb825** [comment]
@yurivict thanks for the extra info!

When comparing your hexdump info for `texture.png`, I think I see what's going on. For reference this is the version in the repository:

```
00000000  89 50 4e 47 0d 0a 1a 0a  00 00 00 0d 49 48 44 52  |.PNG........IHDR|
00000010  00 00 00 04 00 00 00 04  08 02 00 00 00 26 93 09  |.............&..|
00000020  29 00 00 00 09 70 48 59  73 00 00 0b 13 00 00 0b  |)....pHYs.......|
00000030  13 01 00 9a 9c 18 00 00  00 07 74 49 4d 45 07 e1  |..........tIME..|
00000040  01 15 16 04 20 0c c2 d6  94 00 00 00 1d 69 54 58  |.... ........iTX|
00000050  74 43 6f 6d 6d 65 6e 74  00 00 00 00 00 43 72 65  |tComment.....Cre|
00000060  61 74 65 64 20 77 69 74  68 20 47 49 4d 50 64 2e  |ated with GIMPd.|
00000070  65 07 00 00 00 27 49 44  41 54 08 d7 35 c3 b1 0d  |e....'IDAT..5...|
00000080  00 30 0c c3 30 39 8f 37  be 5c 99 4a 80 00 fe 41  |.0..09.7.\.J...A|
00000090  25 44 64 36 a4 c6 96 cc  db b8 b5 80 07 9d 27 11  |%Dd6..........'.|
000000a0  84 38 ef 5f a2 00 00 00  00 49 45 4e 44 ae 42 60  |.8._.....IEND.B`|
000000b0  82                                                |.|
000000b1
```

From this output, one byte is missing from your version of the file. Upon further inspection, and seeing the output when building the port locally, it looks like `dos2unix` is being run on all the files within the repository. When running `dos2unix` manually on that file I can reproduce the same file contents, corrupting the file. (it is no longer a valid PNG)

When looking at [the port file](https://cgit.freebsd.org/ports/tree/misc/cuttlefish/Makefile) I see there's a dependency on `dos2unix`. I don't see any lines that invoke `dos2unix`, does the port system run automatically on all files if this dependency is present? I believe this used to be needed when the `FreeImage` submodule was used, as that repository had some files with DOS line endings. However, now that `graphics/freeimage` is a dependency the submodule will be ignored (linking to the system version instead), so you should be able to remove that dependency from Cuttlefish.

P.S. you should also be able to remove the following, since the working directory should be set to read `texture.png` automatically.

```
pre-test:
	${LN} -s ${WRKSRC}/tool/test/texture.png ${TEST_WRKSRC}/ || true
```

**Comment by yurivict** [comment]
I see. It must have been that some files had Windows newline characters and USES=dos2unux was added there. It is applied to all files, regardless of file type, and it damaged this png file. I just removed it.

--------------------------


Now there are 5 test failures:
```
13/84 Test #13: CuttlefishSaveBC6H ...................***Failed    0.01 sec
Illegal instruction
got exit code 132; expected 0
exit: Illegal number: -1

      Start 14: CuttlefishSaveBC6HUNorm
14/84 Test #14: CuttlefishSaveBC6HUNorm ..............   Passed    0.01 sec
      Start 15: CuttlefishSaveBC6HUFloat
15/84 Test #15: CuttlefishSaveBC6HUFloat .............***Failed    0.01 sec
Illegal instruction
got exit code 132; expected 0
exit: Illegal number: -1

      Start 16: CuttlefishSaveBC6HFloat
16/84 Test #16: CuttlefishSaveBC6HFloat ..............***Failed    0.01 sec
Illegal instruction
got exit code 132; expected 0
exit: Illegal number: -1

      Start 17: CuttlefishSaveETC2
17/84 Test #17: CuttlefishSaveETC2 ...................   Passed    0.01 sec
      Start 18: CuttlefishSaveETC2sRGB
18/84 Test #18: CuttlefishSaveETC2sRGB ...............   Passed    0.01 sec
      Start 19: CuttlefishSaveASTC
19/84 Test #19: CuttlefishSaveASTC ...................   Passed    0.01 sec
      Start 20: CuttlefishSaveASTCHDR
20/84 Test #20: CuttlefishSaveASTCHDR ................   Passed    0.01 sec
      Start 21: CuttlefishSaveASTCsRGB
21/84 Test #21: CuttlefishSaveASTCsRGB ...............   Passed    0.01 sec
      Start 22: CuttlefishSaveASTCLowest
22/84 Test #22: CuttlefishSaveASTCLowest .............   Passed    0.01 sec
      Start 23: CuttlefishSaveASTCLow
23/84 Test #23: CuttlefishSaveASTCLow ................   Passed    0.01 sec
      Start 24: CuttlefishSaveASTCHigh
24/84 Test #24: CuttlefishSaveASTCHigh ...............   Passed    0.01 sec
      Start 25: CuttlefishSaveASTCHightest
25/84 Test #25: CuttlefishSaveASTCHightest ...........   Passed    0.02 sec
      Start 26: CuttlefishSaveB10G11R11_UFloat
26/84 Test #26: CuttlefishSaveB10G11R11_UFloat .......   Passed    0.01 sec
      Start 27: CuttlefishSaveE5B9G9R9_UFloat
27/84 Test #27: CuttlefishSaveE5B9G9R9_UFloat ........   Passed    0.01 sec
      Start 28: CuttlefishSavesRGB
28/84 Test #28: CuttlefishSavesRGB ...................   Passed    0.01 sec
      Start 29: CuttlefishSaveUNorm
29/84 Test #29: CuttlefishSaveUNorm ..................   Passed    0.01 sec
      Start 30: CuttlefishSaveSNorm
30/84 Test #30: CuttlefishSaveSNorm ..................   Passed    0.01 sec
      Start 31: CuttlefishSaveUInt
31/84 Test #31: CuttlefishSaveUInt ...................   Passed    0.01 sec
      Start 32: CuttlefishSaveInt
32/84 Test #32: CuttlefishSaveInt ....................   Passed    0.01 sec
      Start 33: CuttlefishSaveUFloat
33/84 Test #33: CuttlefishSaveUFloat .................   Passed    0.01 sec
      Start 34: CuttlefishSaveFloat
34/84 Test #34: CuttlefishSaveFloat ..................***Failed    0.01 sec
Illegal instruction
got exit code 132; expected 0
exit: Illegal number: -1

```

**Comment by akb825** [comment]
What architecture are you running on? If it's x86 32-bit, I could see this being an issue if you don't have at least SSE2 instructions available. This should be guaranteed for x86 64-bit, as well as most modern processors, but may not be available for older x86 32-bit processors or on VMs if the proper virtualization settings aren't enabled. I might need to add a runtime check for the x86 32-bit target and use the fallback encoders if the required instruction sets aren't available.

**Comment by yurivict** [comment]
This is amd64 (x86_64).

**Comment by akb825** [comment]
I tracked these latest failures down to the hardware accelerated half-float conversion. I had checks to avoid using the hardware acceleration when not available, but I needed to have an extra layer of separation to avoid the illegal instruction exceptions. This will be fixed in the 2.7.3 release that I'm preparing now.

**Comment by akb825** [comment]
I've published the 2.7.3 release, hopefully everything will pass now.

**Comment by yurivict** [comment]
All tests pass now.

Thank you!
